### PR TITLE
observability: implement configuration parsing for observability

### PR DESCRIPTION
**Pull request by sanjaypujare**
Implements g11y observability config parsing required for logging

### CODE REVIEW

**Comment by sanjaypujare** [comment]
friendly ping

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+/** gRPC Observability configuration processor. */
+final class ObservabilityConfig {
+  private static final String CONFIG_ENV_VAR_NAME = "GRPC_CONFIG_OBSERVABILITY";
+
+  private boolean enableCloudLogging = true;
+  private String destinationProjectId = null;
+  private LogFilter[] logFilters;
+  private GrpcLogRecord.EventType[] eventTypes;
+
+  public static class LogFilter {
```
Comment (temawi) [inline observability/src/main/java/io/grpc/observability/ObservabilityConfig.java:38]: Public class should have some JavaDoc explaining what it represents. Same for the fields.

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    public String pattern;
+    public Integer headerBytes;
+    public Integer messageBytes;
+  }
+
+  ObservabilityConfig() {
```
Comment (temawi) [inline observability/src/main/java/io/grpc/observability/ObservabilityConfig.java:44]: I don't think you need this constructor at all.

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+  }
+
+  void read() throws IOException {
```
Comment (temawi) [inline observability/src/main/java/io/grpc/observability/ObservabilityConfig.java:47]: I suggest you make this a static factory method that returns a new ObservabilityConfig instance. That way the client does not need to separately create a new instance and then call read(). Also eliminates the possibility of an instance from being in an intermediate/invalid state where it is created but has no data read into it.

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    parse(System.getenv(CONFIG_ENV_VAR_NAME));
+  }
+
+  @SuppressWarnings("unchecked")
+  void parse(String config) throws IOException {
+    checkArgument(config != null, CONFIG_ENV_VAR_NAME + " value is null!");
+    Map<String, ?> map = (Map<String, ?>) JsonParser.parse(config);
```
Comment (temawi) [inline observability/src/main/java/io/grpc/observability/ObservabilityConfig.java:54]: No need to assign to a variable, you can inline the JsonParse call in the next line.

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+  private void parseLoggingConfig(Map<String,?> loggingConfig) {
+    if (loggingConfig != null) {
+      Boolean value = JsonUtil.getBoolean(loggingConfig, "enable_cloud_logging");
+      if (value != null) {
+        enableCloudLogging = value;
+      }
+      destinationProjectId = JsonUtil.getString(loggingConfig, "destination_project_id");
+      List<?> rawList = JsonUtil.getList(loggingConfig, "log_filters");
+      if (rawList != null) {
+        List<Map<String, ?>> logFilters = JsonUtil.checkObjectList(rawList);
```
Comment (temawi) [inline observability/src/main/java/io/grpc/observability/ObservabilityConfig.java:68]: nit: the naming here is a bit confusing as it's the same as the logFilters field at the class level. Maybe call this jsonLogFilters or something to make it clear they are not of the same type?

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+        List<String> enumTypes = JsonUtil.checkStringList(rawList);
+        this.eventTypes = new GrpcLogRecord.EventType[enumTypes.size()];
+        for (int i = 0; i < enumTypes.size(); i++) {
+          this.eventTypes[i] = convert(enumTypes.get(i));
+        }
+      }
+    }
+  }
+
+  private GrpcLogRecord.EventType convert(String val) {
```
Comment (temawi) [inline observability/src/main/java/io/grpc/observability/ObservabilityConfig.java:85]: nit: call this convertEventType?

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java (hunk 1 above)
Comment (sanjaypujare) [reply observability/src/main/java/io/grpc/observability/ObservabilityConfig.java:38]: Yes, I need to add a few more things, such as final etc . Will do

File: observability/src/main/java/io/grpc/observability/ObservabilityConfig.java [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+      case "GRPC_CALL_HALF_CLOSE":
+        return GrpcLogRecord.EventType.GRPC_CALL_HALF_CLOSE;
+      case "GRPC_CALL_CANCEL":
+        return GrpcLogRecord.EventType.GRPC_CALL_CANCEL;
+      default:
+        throw new IllegalArgumentException("Unknown event type value:" + val);
+    }
+  }
+
+  private LogFilter parseLogFilter(Map<String,?> map) {
```
[the remark this answers is not in the document]
Comment (sanjaypujare) [reply observability/src/main/java/io/grpc/observability/ObservabilityConfig.java]: Did you mean the variable (parameter) `map` or the method name? I misread it as talking about the method name.

[the remark this answers is not in the document]
Comment (temawi) [reply observability/src/main/java/io/grpc/observability/ObservabilityConfig.java]: I mean the variable name, "map". Just saying that simply using the type of the object as the variable name is not the most descriptive convention.

[the remark this answers is not in the document]
Comment (sanjaypujare) [reply observability/src/main/java/io/grpc/observability/ObservabilityConfig.java]: yes, I figured. It is addressed by the way.

**Review by temawi — APPROVED** [review]

**Pull request merged**